geps/gep-1867.md

```diff
@@ -0,0 +1,128 @@
+This has been previously discussed in [this issue](https://github.com/kubernetes-sigs/gateway-api/issues/567).
+
+As a cluster scoped resource, `GatewayClass` does not meet this requirement.
+This restricts customization use cases to either a few pre-provisioned classes by the admin, or running in an environment where the "Infrastructure Provider" and "Cluster Operator" are the same roles.
+The distinction between these roles is explicitly called out on the homepage: https://gateway-api.sigs.k8s.io/#what-is-the-gateway-api.
+
+### Custom Resource
+
+`parametersRef` is entirely a generic implementation-specific meaning.
+This means implementations will either need a custom CRD or use untyped resources like ConfigMap.
+Neither of these have any consistency between implementations.
+While there will always be some vendor-specific requirements, there are also a number of configuration aspects of a Gateway that are common between implementations.
+However, these cannot currently be expressed in a vendor-neutral way.
+
+Additionally, there is hesitancy to use a CRD (which leads to CRD proliferation), which pushes users towards untyped ConfigMaps which are not much better than annotations. The scoping, as mentioned above, is also a bit awkward of a cluster scoped resource pointing to a namespaced object.
+
+### Separation of concerns
+
+While there is value out of providing class-wide options as defaults, there is also value in providing these options on the object (Gateway) directly.
+
+Some parallels in existing APIs:
+
+[Policy Attachment](https://gateway-api.sigs.k8s.io/references/policy-attachment) offers a hierarchy of defaults and overrides, allowing attachment to GatewayClass and Gateway.
+This is similar to our needs here, but representing infrastructure configuration as a "Policy" is a bit problematic, and the existing mechanisms have no hierarchy.
+
+In core Kubernetes, Pods declare their requirements (for example, CPU requests) inline in the Pod resource; there is not a `ResourceClass` API that abstracts these further.
+These higher level abstractions are handled by layered APIs (whether this is a CRD, an admission webhook, CI/CD tooling, etc).
+This allows users the flexibility to easily configure things per-pod basis.
+If the infrastructure admin wants to impose defaults or requirements on this flexibility, they are able to do so (in fact, `LimitRanger` provides a built in mechanism to do so).
+
+### Dynamic Changes
+
+Currently, the spec recommends `GatewayClass` to be used as a *template*.
+Changes to it are not expected to change deployed `Gateway`s.
+
+This makes usage problematic in a declarative way.
+For example, if I wanted to represent a `version` field and change that to trigger an upgrade, I would need to create an entirely new `Gateway`.
```

Comment:
I think in most cases it's significantly safer to spin up a new Gateway when doing any kind of meaningful infra-level changes, but I'd agree that it's overkill for bumping the image version used. There are some simpler ways to solve this though:

1. Suggest that implementations document which fields will result in in-place upgrades vs only apply to future Gateways
2. Allow Gateway owners to specify if they want to subscribe to automatic updates from the GatewayClass

Reply:
Maybe it would help to retitle and refocus the current "Challenges with GatewayClass" section to "Why not GatewayClass params?" or similar. I don't think this GEP is saying that we shouldn't use GatewayClass at all, just that there are some cases where Gateway level config would be helpful. I think this GEP should focus on the specific situations where Gateway-level config is more useful than GWC-level config. 

I'm not even sure that either of my suggestions above need to be alternatives considered. I think either or both could be reasonable extensions of the GEP without significantly changing it.

I think this is really getting to an overall theme of my comments/perspective on this GEP: We need principles for when to introduce infra config at each level. I don't think every GWC-level param should have a GW-level equivalent, and of course the same is true in the other direction.

Reply:
I updated this to provide principals